extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleSetTableOwner.scala

```diff
@@ -0,0 +1,110 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.plugin.spark.authz.ranger
+
+import org.apache.spark.sql.{Row, SparkSession}
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.analysis.{PersistedView, ViewType}
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.catalyst.plans.logical.{Command, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.command.RunnableCommand
+
+import org.apache.kyuubi.plugin.spark.authz.util.AuthZUtils._
+import org.apache.kyuubi.plugin.spark.authz.util.WithInternalChildren
+
+class RuleSetTableOwner(spark: SparkSession) extends Rule[LogicalPlan] {
+  import RuleSetTableOwner._
+
+  private val setOwnerEnabled =
+    spark.conf.get(SET_OWNER_ENABLED_KEY, SET_OWNER_ENABLED_DEFAULT).toLowerCase == "true"
+  private val authzUser = getAuthzUgi(spark.sparkContext).getShortUserName
+  private val catalog = spark.sessionState.catalog
+
+  private val commandToTableField = Map(
+    "CreateDataSourceTableCommand" -> "table",
+    "CreateDataSourceTableAsSelectCommand" -> "table",
+    "CreateTableCommand" -> "table",
+    "CreateHiveTableAsSelectCommand" -> "tableDesc",
+    "OptimizedCreateHiveTableAsSelectCommand" -> "tableDesc")
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!setOwnerEnabled) {
+      return plan
+    }
+
+    plan match {
+      case c: Command if commandToTableField.contains(c.nodeName) =>
+        setTableOwner(plan, commandToTableField(c.nodeName))
+        plan
+      case r: RunnableCommand if r.nodeName == "CreateTableLikeCommand" =>
+        val tableId = getFieldVal[TableIdentifier](r, "targetTable")
+        if (!catalog.tableExists(tableId)) {
+          CreateTableAndSetOwnerCommand(tableId, authzUser, r)
+        } else r
+      case r: RunnableCommand
+          if r.nodeName == "CreateViewCommand" && getFieldVal[ViewType](
+            r,
+            "viewType") == PersistedView =>
+        val tableId = getFieldVal[TableIdentifier](r, "name")
+        val replace = getFieldVal[Boolean](r, "replace")
+        if (replace || !catalog.tableExists(tableId)) {
+          CreateTableAndSetOwnerCommand(tableId, authzUser, r)
+        } else r
+      case _ =>
+        plan
+    }
+  }
+
+  private def setTableOwner(plan: LogicalPlan, catalogTableField: String): Unit = {
+    val table = getFieldVal[CatalogTable](plan, catalogTableField)
+    try {
+      val field = plan.getClass.getDeclaredField(catalogTableField)
+      field.setAccessible(true)
+      field.set(plan, table.copy(owner = authzUser))
```

Comment:
Let's generalize them to a new method `setFieldVal` in `AuthzUtils` ?

Reply:
Done